### ISSUE TITLE
time or distance-optimal route plans have poor runtime performance

### ISSUE THREAD

**Issue by robfitzgerald** [NREL/routee-compass#92]
after a monster refactor to support parameterized cost models, we were getting normal energy-optimal runtimes from the EnergyTraversalModel, but time and distance-optimal routes are still taking long, suggesting that the a* heuristic is not correctly calculated for those objectives. see table [here](https://github.com/NREL/routee-compass/pull/87#discussion_r1435289289).

to reproduce, use this test case on our internal tomtom-based denver scenario:

```json
{
  "model_name": "2017_CHEVROLET_Bolt",
  "starting_soc_percent": 100,
  "destination_y": 39.62627481432341,
  "destination_x": -104.99460207519721,
  "origin_y": 39.798311884359094,
  "origin_x": -104.86796368632217,
  "state_variable_coefficients": {
    "distance": 0,
    "time": 0,
    "energy_electric": 0
  }
}
```

**Comment by nreinicke** [comment]
Did a bit of digging into this today. It appears that we might be able to gain some runtime improvements by tweaking the heuristic functions a bit. 

## TLDR;
It could make sense to expose some control of the heuristic function to the configuration so we can adjust based on the desired results and runtimes.

For example, if we consider the distance heuristic as the haversine distance between the search point and the target point, we're guaranteed to never overestimate that cost, but, in reality the network will likely display some [circuity](https://en.wikibooks.org/wiki/Transportation_Geography_and_Network_Science/Circuity). So we could try to bit a bit less optimistic for our distance heuristic and apply a circuity factor. Here's are a couple of examples of using a circuity of 1.0 (baseline) and 1.2:

## circuity of 1.0
![image](https://github.com/NREL/routee-compass/assets/1743744/cd935b86-dc62-4dd2-a161-10089adbfd64)

<details>

<summary>Routes</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/b6c8041a-af5e-4a24-867d-80efe0367320)

</details>

<details>

<summary>Distance Tree</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/8b0f819d-84a1-45aa-8bb1-b46197c04dc3)

</details>

<details>

<summary>Time Tree</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/6a3a8bd0-41e8-43ce-bb5d-490864853606)

</details>

<details>

<summary>Energy Tree</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/e9718e6b-3550-4f22-8d48-d0e9dc444e2c)

</details>



## circuity of 1.2
![image](https://github.com/NREL/routee-compass/assets/1743744/4addf5fd-8192-4e60-84d9-1cd00a1821fd)

<details>

<summary>Routes</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/46a01d75-7b5a-4faf-8688-f4fe373b1763)

</details>

<details>

<summary>Distance Tree</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/36abac3c-1f2b-4247-baa6-87a802bcc347)

</details>

<details>

<summary>Time Tree</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/fc31f439-282b-4bea-9bf8-d39d837cc2bd)

</details>

<details>

<summary>Energy Tree</summary>

![image](https://github.com/NREL/routee-compass/assets/1743744/a8132ac8-e68d-419c-9bf8-9e2c48d8d23e)

</details>



The tradeoff here is that applying the circuity factor of 1.2 results in smaller tree and improved runtime (with the exception of the energy only case, _what the heck is happening there??_) but the circuity factor of 1.0 finds marginal improvements to the distance and energy shortest paths.

**Comment by robfitzgerald** [comment]
> So we could try to bit a bit less optimistic for our distance heuristic and apply a circuity factor. Here's are a couple of examples of using a circuity of 1.0 (baseline) and 1.2:

are you multiplying the cost by the circuitry factor? i assume that this would tighten the estimate for _this exact instance_ of the best distance and time routes, since they in fact are bowed out (circuitous). but i'm guessing it didn't help the energy case, since the actual optimal path there is fairly diagonal (has a low circuity that we may be over-estimating with a value of 1.2). my instinct is that this would not generalize to cases where the opposite is true, but i honestly haven't done my homework on improvements to a-star heuristics to be certain. 

speaking of, if you are content that the code seems correct, then maybe our next step here is to look at how people go "beyond the basics" with a-star to address performance for more complex cost functions, or to investigate how [inconsistent](https://en.wikipedia.org/wiki/Consistent_heuristic) our energy-optimal cost function becomes.

> with the exception of the energy only case, what the heck is happening there??

i would love to nod in agreement here, but, it appears that time + energy maps for the 1.2 case are the same map. guessing it's the time map just got uploaded twice? :sweat_smile:

**Comment by nreinicke** [comment]
> guessing it's the time map just got uploaded twice? 

🤦 yep, just fixed, that's what I get for moving fast. 

The weirdness I'm commenting on is that in the first case, the energy tree is 140k with a runtime of 23 seconds but then in the second case, the energy tree is less than half the size, about 65k but the runtime goes up to 63 seconds... 

> are you multiplying the cost estimate by the circuitry factor? 

In this case, I was multiplying the raw distance estimate by the circuity factor which gets propagated to the cost estimates for time and energy. It is a good point that this heuristic adjustment might be best suited to the time case since both energy and distance optimal routes likely take a route similar to the straight line distance. 

> then maybe our next step here is to look at how people go "beyond the basics" with a-star to address performance for more complex cost functions, 

Yeah that seems like a good idea.

**Comment by robfitzgerald** [comment]
> in the first case, the energy tree is 140k with a runtime of 23 seconds but then in the second case, the energy tree is less than half the size, about 65k but the runtime goes up to 63 seconds...

i think we are missing an important metric here to make sense of it, which is the number of search iterations. as opposed to dijkstra's, it is possible for a-star to revisit the same link multiple times to update the cost value. so, with the 63 second run, it is likely revisiting links many many times. something we should probably add to the traversal summary, as well as the average link visits value (`iterations / tree_size`).

> I was multiplying the raw distance estimate by the circuity factor which gets propagated to the cost estimates for time and energy

right. but while this circuity distance heuristic is helpful for demonstration, we can't really use it, right? it would require knowing the circuity of the route we are searching for ahead of time, or the average circuity of the network that will be used for a given search. in our case, where we often have the whole nation loaded, we can't easily pre-compute this value for incoming searches, and the national average value would be wrong in many places (like, in phoenix vs boston, that value would be drastically different because of the differences in topology).

**Comment by nreinicke** [comment]
> i think we are missing an important metric here to make sense of it, which is the number of search iterations. as opposed to dijkstra's, it is possible for a-star to revisit the same link multiple times to update the cost value.

right, that's a really good point and it would be great to see why this is happening, maybe related to zero cost links?

> while this circuity distance heuristic is helpful for demonstration, we can't really use it, right? 

yeah I think I agree with that since, as you describe, we can't actually guarantee that it will be admissible for all cases.

> maybe this [bev zero energy] limit value is causing issues in a way i'm not realizing?

yeah that's a good question that probably merits further digging... 

Thanks for the thoughtful response here! I've been focusing on an idea to speed up the energy prediction model that is a bit more robust than caching values. I'll share more next week!

**Comment by robfitzgerald** [comment]
i have a hunch that (#111) may have, or (#115) might fix this, since both will standardize CPU + allocations around cost modeling and state updates, respectively.

**Comment by nreinicke** [comment]
I just re-ran the sample query that I shared above:

![image](https://github.com/NREL/routee-compass/assets/1743744/04b0aa68-48ab-4bc6-807a-a7a959125f5a)

Based on our previous runs, it looks like runtime has come down about 10x which is exciting.
